prereise/gather/demanddata/transportation_electrification/immediate.py

```diff
@@ -0,0 +1,191 @@
+):
+    """Immediate charging function
+
+    :param int census_region: any of the 9 census regions defined by US census bureau.
+    :param int model_year: year that is being modelled/projected to, 2017, 2030, 2040, 2050.
+    :param int veh_range: 100, 200, or 300, represents how far vehicle can travel on single charge.
+    :param int kwhmi: fuel efficiency, should vary based on vehicle type and model_year.
+    :param int power: charger power, EVSE kW.
+    :param int location_strategy: where the vehicle can charge-1, 2, or 3; 1 = home only, 2 = home and work, 3 = everywhere
+    :return: (*numpy.ndarray*) -- charging profiles.
+    """
+    # load NHTS data from function
+    newdata = mileage.remove_ldt(mileage.load_data(census_region))
+    # add new columns to newdata to store data that is not in NHTS data
+    newdata['trip start battery charge'] = 0
+    newdata['trip end battery charge'] = 0
+    newdata['charging power'] = 0
+    newdata['charging time'] = 0
+    # jx: never actually used... delete?
+    newdata['charging consumption'] = 0  # needs double check
+    newdata['BEV could be used'] = 0
+    newdata['trip number'] = 0
+
+    input_day = mileage.get_input_day(mileage.get_model_year_dti(model_year))
+
+    TRANS_charge = np.zeros(24 * len(input_day))
+
+    data_day = mileage.get_data_day(newdata)
+
+    actual_vmt = 0
+    potential_vmt = 0
+
+    # charging when SOC less than batterstrategy
+    battery_strategy = 1
+
+    # 1-anytrip number, 2-last trip
+    trip_strategy = 1
+
+    kwh = kwhmi * veh_range
+
+    for day_iter in range(len(input_day)):
+
+        electricload = np.zeros(4800)
+        # initializes electricload_old
+
+        # vmt, trip amount, vehicle amount
+        info1 = [0] * 3
+
+        # meh: better names for counters
+        trip_num = 1  # trip number for current vehicle
+        # charging start and end points
+        start_point = 0
+        end_point = 0
+
+        # flag to see if the trip is the first for the vheicle, '1' yes, '0' no
+        firstrip = 0
+
+        # meh: day2 = data_day, month2 = data_month
+        for i in range(len(newdata)):
+            if data_day[i] == input_day[day_iter]:
+                if (
+                    i > 0
+                    and newdata.loc[i]["sample vehicle number"]
+                    == newdata.loc[i - 1]["sample vehicle number"]
+                ):
+                    newdata.loc[i]["trip start battery charge"] = (
+                        newdata.loc[i - 1]["trip end battery charge"]
+                        + newdata.loc[i - 1]["depleting time"]
+                    )
+                    # trip number
+                    trip_num += 1
+                    newdata.loc[i]["trip number"] = trip_num
+                else:
+                    newdata.loc[i]["trip start battery charge"] = kwh
+                    trip_num = 1
+                    # trip number
+                    newdata.loc[i]["trip number"] = trip_num

```

Comment:
It seems that most of what we're doing here is assigning trip numbers based on grouping by vehicle number, and then assigning `"trip start battery charge"` based on the value within the trip number. The former can be accomplished outside of a loop using something like `newdata.loc["trip number"] = newdata.groupby("sample vehicle number").rank(method="first")`.

Reply:
Is there an implicit assumption here that the data entries are in chronological order?